### DISCUSSION TITLE
uhubctl watchdog

### DISCUSSION THREAD

**Discussion by WirthmU** [mvp/uhubctl#583] [Ideas]
ChatGPT have me this:

To create a USB watchdog script for a Raspberry Pi using `uhubctl`, follow these steps:

### Step 1: Install uhubctl
First, you need to install `uhubctl`, which allows you to control the power of USB ports.

1. Update your Raspberry Pi:
   ```bash
   sudo apt update
   sudo apt upgrade
   ```

2. Install `uhubctl`:
   ```bash
   sudo apt install uhubctl
   ```

### Step 2: Identify Your USB Hub and Port

Use `uhubctl` to list your USB hubs and ports:
```bash
uhubctl
```
This will give you information about available USB hubs and ports, which you’ll need to control specific ports.

### Step 3: Create the Watchdog Script

You can create a script that monitors a particular service or process and toggles the power on a USB port if the service/process is unresponsive.

Here’s an example script:

```bash
#!/bin/bash

# Configuration
SERVICE="your-service-name"    # Replace with the name of the service/process to monitor
HUB=1                          # Replace with the correct hub number from uhubctl
PORT=2                         # Replace with the correct port number from uhubctl

# Function to check if the service is running
check_service() {
    if systemctl is-active --quiet $SERVICE; then
        echo "$SERVICE is running"
        return 0
    else
        echo "$SERVICE is not running"
        return 1
    fi
}

# Function to reset the USB port
reset_usb_port() {
    echo "Resetting USB port $PORT on hub $HUB"
    uhubctl -l $HUB -p $PORT -a off
    sleep 2
    uhubctl -l $HUB -p $PORT -a on
}

# Main loop
while true; do
    if ! check_service; then
        echo "Service is down. Restarting USB port..."
        reset_usb_port
        echo "Sleeping for 60 seconds before next check..."
        sleep 60  # Wait 60 seconds before checking again
    else
        echo "Service is running fine. Sleeping for 10 seconds..."
        sleep 10  # Check every 10 seconds
    fi
done
```

### Step 4: Make the Script Executable
Give execution permission to the script:
```bash
chmod +x /path/to/your-script.sh
```

### Step 5: Run the Script
You can run the script manually or set it up to run on startup.

To run the script manually:
```bash
/path/to/your-script.sh
```

### Step 6: (Optional) Run the Script on Startup
To run the script on startup, you can add it to `/etc/rc.local` or create a systemd service.

For `/etc/rc.local`, add the following line before `exit 0`:
```bash
/path/to/your-script.sh &
```

Or, you can create a systemd service file:
1. Create a new service file:
   ```bash
   sudo nano /etc/systemd/system/usb-watchdog.service
   ```

2. Add the following content:

   ```ini
   [Unit]
   Description=USB Watchdog Service
   After=network.target

   [Service]
   ExecStart=/path/to/your-script.sh
   Restart=always

   [Install]
   WantedBy=multi-user.target
   ```

3. Enable and start the service:
   ```bash
   sudo systemctl enable usb-watchdog.service
   sudo systemctl start usb-watchdog.service
   ```

This script will monitor your specified service and reset the USB port if the service is down, effectively acting as a watchdog for connected devices.

Anyone thinks this is feasible on a RPI4?

**Comment by mvp** [comment]
I don't see straight connection between state of the watched service and state of USB ports. If service is down, it doesn't necessarily mean that anything was wrong with USB. And if service is not running, this script doesn't do anything about it (e.g. try to start it again), but simply resets USB ports. Long story short, I wouldn't necessarily trust anything that chatgpt recommends.

**Reply by WirthmU** [reply]
Well at least it's working by hand.

If I see the allsky service, which is what I'm using to drive an USB-Allsky camera (https://github.com/AllskyTeam), is down on the PI, and then I do reset the USB port using uhubctl commands:

`uhubctl -l 1-1 -p 1 -a off`

and then about 30 seconds later

`uhubctl -l 1-1 -p 1 -a on`

the allsky service and current daytime picture immediately came back.

**Reply by mvp** [reply]
You can certainly create such watchdog to detect if your USB device is misbehaving and automatically turn USB on or off as necessary. However, it is not always related to the state of any particular service.

E.g. some users implemented similar automatic uhubctl USB on/off script to automatcally turn USB fan on if CPU temperature is above certain threshold, and turn it off it temperature drops back below safe level.

uhubctl is a basic tool for you to build any workflow you want - react to certain events and turn USB off and on. It is up to you how you build your workflow and what you react upon.

I recommend reading [notable projects list](https://github.com/mvp/uhubctl?tab=readme-ov-file#notable-projects-using-uhubctl) for great examples.
